### ISSUE TITLE
istio-cni documentation issues

### ISSUE THREAD

**Issue by sdake** [istio/istio.io#4327]
As a new consumer of CNI, I took notes during my first interaction with the system.  Looks pretty good!  The docs could be refined as follows:

1. Where is the step 3 tarball?  I cloned the repo to obtain the `deployments` directory which contains the helm templates - but this is the wrong approach.
1. The install has a cut and paste, but is missing the crd install step, which may confuse people (it confused me on first attempting this work…)
1. Remove the word ‘Example:”
1. Don’t we always wish to exclude kube-system and kube-public?
1. Reword from: `Not all hosted Kubernetes clusters are created with the kubelet configured to use the CNI plugin so compatibility with this istio-cni solution is not ubiquitous`. Reword to:
`The istio-cni solution is not ubiquitous.  Some platforms, especially hosted Kubernetes environments, do not enable the CNI plugin in the kubelet configuration.`

1. The first row in the hosted table list is confusing.  As I run on bare metal (which is obviously not a hosted platform) - I was misled into thinking that first row applied to my use case.   I’d rework the table into 4  columns.  rename “Required non-default settings” to “Configuration Overrides” and add a new field “Platform Overrides”.  For GKE I’d put “Enable Network Policy.  This allows you to delete the first row entirely.  If no overrides are required, make sure to put the word “none” for the override.  An empty table field could mean many things to many people :)

1. Remove single bullet: https://preliminary.istio.io/docs/setup/kubernetes/additional-setup/cni/#gke-setup.  Change the single bullet into a preceding warning using {{< warning >}} - single bullets don’t make any sense.

1. What does This refer to?  The preceding section?  A little confusing..
```
This should have no effect on the operations performed by the base CNI plugin configuring the pod’s networking setup, although not all CNI’s have been validated.
```

We want people to be able to burn through these docs, not spend 30 minutes (as I did) on first exposure to them.  If they want details, cool, they can read them more slowly.

**Comment by john-a-joyce** [comment]
Regarding the tarball.  That should get a different issue because we need to get the tar ball built and decide where it is pushed, before we can fix the documentation side of things

**Comment by tiswanso** [comment]
The Istio release process toolchain seemed to change and no longer create the istio-cni-*.tgz in the release archive.  Maria created this fix but it looks like it never made it:  https://github.com/istio/istio/pull/13615

**Comment by tiswanso** [comment]
@sdake -- thanks for tracking this stuff.  Some Q's/comments for clarification:

a)  For item 2, I'm wary of recreating instructions for installing Istio here.  The paragraph links to the Istio install instructions and this bullet is intended to show setting the parameter `--set istio_cni.enabled=true`
b)  For item 3, you're referring to this heading?  https://preliminary.istio.io/docs/setup/kubernetes/additional-setup/cni/#example-excluding-specific-kubernetes-namespaces
c)  For item 4, it's fine either way to exclude those namespaces or not.  The istio-cni only does something when it sees and istio-proxy container on a pod with more than 1 container (there's more to it but that's the gist).  Excluding at the ns level would prevent the istio-cni from even trying to get more pod spec info from k8s api so there would be some gains in efficiency during pod creates.
d) For item 8, the statement is in regards to compatibility with other CNIs.  It refers to the impact of the istio-cni's changes the pod network ns.

**Comment by sdake** [comment]
> @sdake -- thanks for tracking this stuff. Some Q's/comments for clarification:
> 
> a) For item 2, I'm wary of recreating instructions for installing Istio here. The paragraph links to the Istio install instructions and this bullet is intended to show setting the parameter `--set istio_cni.enabled=true`

Understood, however, there is step 1 to create the namespace.  Then step 2 is install istio.  I ran it 3 or 4 times in cut and paste mode, before it triggering in my mind `Oh, need to install the crds`.  Another option is to completely remove step 1, and change step 2 to something like

`Use the instructions from (installing with helm) with the following additional flag `--set cni.enabled=true`.  See this page for an example of how I'd like to see all install related flags handled:

https://preliminary.istio.io/blog/2018/aws-nlb/#generate-the-istio-manifest

Even this blog post needs a little refinement - would be nice to have a consistent way to document these flags - possibly with a boilerplate.

> b) For item 3, you're referring to this heading? https://preliminary.istio.io/docs/setup/kubernetes/additional-setup/cni/#example-excluding-specific-kubernetes-namespaces

yes - it seems like it would be more in line with a task (or a set of CNI tasks) to be refactored when time permits into istio.io.  In the meantime, no sense adding the extra word.

> c) For item 4, it's fine either way to exclude those namespaces or not. The istio-cni only does something when it sees and istio-proxy container on a pod with more than 1 container (there's more to it but that's the gist). Excluding at the ns level would prevent the istio-cni from even trying to get more pod spec info from k8s api so there would be some gains in efficiency during pod creates.

cool - nm then - no sense churning for effeciency this late in the dev cycle.

> d) For item 8, the statement is in regards to compatibility with other CNIs. It refers to the impact of the istio-cni's changes the pod network ns.

What I meant is the pronoun `This` is used without a reference to what `This` is.  Perhaps you can reword?

**Comment by sdake** [comment]
> Regarding the tarball. That should get a different issue because we need to get the tar ball built and decide where it is pushed, before we can fix the documentation side of things

Agreed.  Filed issue here: https://github.com/istio/istio/issues/14742

**Comment by sdake** [comment]
CNI should be in the next daily release: https://github.com/istio/istio/pull/14752

**Comment by rlenglet** [comment]
The doc misses the installation of the `istio-init` chart. Generally, it should be a copy of the [generic Helm install docs](https://preliminary.istio.io/docs/setup/kubernetes/install/helm/), with additional instructions.

**Comment by howardjohn** [comment]
Published in https://gcsweb.istio.io/gcs/istio-prerelease/daily-build/release-1.2-latest-daily/. Not in rc.1 though.

So we should refactor the docs for how it is structured there, or if needed change how we package the release

**Comment by rlenglet** [comment]
I started an update in #4398. One issue is that to make the instructions complete, there's a lot to copy from the standard Helm instructions. One way to solve this problem would be to:
* Remove the CNI-specific instructions page, and move it into a reference page with no instructions in there.
* Merge the few CNI-specific instructions into the main Helm-based install page, as optional steps, and just refer to it from the CNI reference page.
Does that sound like a good idea?

**Comment by geeknoid** [comment]
Can we address this issue today? This is a P0 blocking the 1.2 release which is planned for tomorrow.

Alternatively, is this really a P0? What state is CNI in (Alpha, Beta, Stable)? Is it a big feature of 1.2?

**Comment by rlenglet** [comment]
@geeknoid Yes, we're doing the last rounds of reviews. Should hopefully be complete shortly.
